src/System.Console/src/System/Console.cs

```diff
@@ -45,6 +50,70 @@ public static TextReader In
             }
         }
 
+        public static Encoding InputEncoding
+        {
+            get
+            {
+                return Volatile.Read(ref _inputEncoding) ?? EnsureInitialized(ref _inputEncoding, () =>
+                {
+                    return ConsolePal.InputEncoding;
+                });
+            }
+            set
+            {
+                CheckNonNull(value, "value");
+                lock (InternalSyncObject)
+                {
+                    // Set the terminal console encoding.
+                    ConsolePal.SetConsoleInputEncoding(value);
+
+                    Volatile.Write(ref _inputEncoding, value);
+
+                    // We need to reinitialize Console.In in the next call to _in
+                    // This will discard the current StreamReader, potentially 
+                    // losing buffered data
+                    Volatile.Write(ref _in, null);
+                }
+            }
+        }
+
+        public static Encoding OutputEncoding
+        {
+            get
+            {
+                return Volatile.Read(ref _outputEncoding) ?? EnsureInitialized(ref _outputEncoding, () =>
+                {
+                    return ConsolePal.OutputEncoding;
+                });
+            }
+            set
+            {
+                CheckNonNull(value, "value");
+
+                lock (InternalSyncObject)
+                {
+                    // Set the terminal console encoding.
+                    ConsolePal.SetConsoleOutputEncoding(value);
+
+                    // Before changing the code page we need to flush the data 
+                    // if Out hasn't been redirected. Also, have the next call to  
+                    // _out reinitialize the console code page.
+                    if (Volatile.Read(ref _out) != null && !_isOutTextWriterRedirected)
+                    {
+                        _out.Flush();
```

Comment:
This will take a different lock, right?  We need to be very careful that there are no code paths where that lock may be held while then taking the lock on InternalSyncObject, as that could result in deadlock.